### PR TITLE
feature[next]: Temporary extraction heuristics

### PR DESCRIPTION
**Pull request by tehrengruber**
Adds a heuristics that only extracts a temporary if the respective lift expr is derefed in more than one position. This should give reasonably good performance and avoids many unnecessary temporaries.

### CODE REVIEW

File: tests/next_tests/unit_tests/iterator_tests/transforms_tests/test_global_tmps.py [hunk 1]
```diff
@@ -31,53 +32,23 @@ def test_split_closures():
     testee = ir.FencilDefinition(
```
Comment (tehrengruber) [inline tests/next_tests/unit_tests/iterator_tests/transforms_tests/test_global_tmps.py:32]: Changes in this function are merely a refactoring to use `ir.makers`.

File: src/gt4py/next/program_processors/runners/gtfn_cpu.py [hunk 2]
```diff
@@ -134,6 +134,9 @@ def compilation_hash(otf_closure: stages.ProgramCall) -> int:
 ](
     name="run_gtfn_with_temporaries",
     otf_workflow=run_gtfn.otf_workflow.replace(
-        translation=run_gtfn.otf_workflow.translation.replace(lift_mode=LiftMode.FORCE_TEMPORARIES),
+        translation=run_gtfn.otf_workflow.translation.replace(
```
Comment (tehrengruber) [inline src/gt4py/next/program_processors/runners/gtfn_cpu.py:137]: Temporary extraction without the heuristics is not useful in almost all cases. I just switched to using the heuristics by default.

File: src/gt4py/next/iterator/transforms/global_tmps.py [hunk 3]
```diff
@@ -483,19 +530,25 @@ def update_domains(node: FencilWithTemporaries, offset_provider: Mapping[str, An
                         nbt_provider = offset_provider[offset_name]
                         old_axis = nbt_provider.origin_axis.value
                         new_axis = nbt_provider.neighbor_axis.value
-                        consumed_domain.ranges.pop(old_axis)
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/global_tmps.py:486]: Unrelated to this PR. This approach of popping from the dict failed when the domain had more than one axis as the order was not preserved.

Comment (havogt) [reply src/gt4py/next/iterator/transforms/global_tmps.py:486]: why does the order matter?

File: src/gt4py/next/iterator/type_inference.py [hunk 4]
```diff
@@ -931,6 +932,9 @@ def visit_StencilClosure(
         )
         return Closure(output=output, inputs=Tuple.from_elems(*inputs))
 
+    def visit_FencilWithTemporaries(self, node: FencilWithTemporaries, **kwargs):
+        return self.visit(node.fencil, **kwargs)
+
```
Comment (havogt) [inline src/gt4py/next/iterator/type_inference.py:937]: do we need this?

**Review by havogt — CHANGES_REQUESTED** [review]

File: src/gt4py/next/iterator/transforms/global_tmps.py [hunk 5]
```diff
@@ -445,7 +492,12 @@ def update_domains(node: FencilWithTemporaries, offset_provider: Mapping[str, An
         if closure.domain == AUTO_DOMAIN:
             # every closure with auto domain should have a single out field
             assert isinstance(closure.output, ir.SymRef)
+
+            if closure.output.id not in domains:
+                raise NotImplementedError(f"Closure output {closure.output.id} is never used.")
```
Comment (DropD) [inline src/gt4py/next/iterator/transforms/global_tmps.py:497]:
```suggestion
                raise NotImplementedError(f"Closure output '{closure.output.id}' is never used.")
```

File: src/gt4py/next/iterator/transforms/pass_manager.py [hunk 6]
```diff
@@ -118,6 +122,32 @@ def apply_common_transforms(
...
+            inlined = InlineLambdas.apply(
+                inlined,
+                opcount_preserving=True,
+                force_inline_lift_args=True,
+            )
+            if inlined == ir:
+                break
+            ir = inlined
+        else:
+            raise RuntimeError("Inlining lift and lambdas did not converge.")
```
Comment (DropD) [inline src/gt4py/next/iterator/transforms/pass_manager.py:144]:
```suggestion
            raise RuntimeError("Inlining 'lift' and 'lambdas' did not converge.")
```

**Review by DropD — COMMENTED** [review]
Small changes to conform to upcoming error message guidelines.

File: src/gt4py/next/iterator/transforms/global_tmps.py [hunk 7]
```diff
@@ -512,13 +561,22 @@ def update_domains(
                             (axis, range_) if axis != old_axis else (new_axis, new_range)
                             for axis, range_ in consumed_domain.ranges.items()
                         )
+                        # TODO(tehrengruber): Revisit. Somehow the order matters so preserve it.
```
Comment (havogt) [inline src/gt4py/next/iterator/transforms/global_tmps.py:564]: for `unstructured_domain(horizontal, vertical)` the order matters, is that the point here?

**Review by havogt — APPROVED** [review]
One comment about your TODO, maybe you want to update it.

**Comment by tehrengruber** [comment]
cscs-ci run

File: src/gt4py/next/iterator/transforms/global_tmps.py (hunk 7 above)
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/global_tmps.py:564]: I'm guessing that it is more the `TemporaryAllocation` where the order matters, but I am not sure.

**Pull request merged**